tests/interfaces/test_batch_tensorflow_qnode.py

```diff
@@ -0,0 +1,1057 @@
+from pennylane.beta import qnode, QNode
+from pennylane.tape import JacobianTape
+
+
+qubit_device_and_diff_method = [
+    ["default.qubit", "finite-diff", "backward"],
+    ["default.qubit", "parameter-shift", "backward"],
+    ["default.qubit", "backprop", "forward"],
+    ["default.qubit", "adjoint", "forward"],
+    ["default.qubit", "adjoint", "backward"],
+]
+
+
+@pytest.mark.parametrize("dev_name,diff_method,mode", qubit_device_and_diff_method)
+class TestQNode:
+    """Test that using the QNode with TensorFlow integrates with the PennyLane stack"""
+
+    def test_execution_with_interface(self, dev_name, diff_method, mode):
+        """Test execution works with the interface"""
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, interface="tf", diff_method=diff_method, mode=mode)
+        def circuit(a):
+            qml.RY(a, wires=0)
+            qml.RX(0.2, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        a = tf.Variable(0.1)
+        circuit(a)
+
+        # if executing outside a gradient tape, the number of trainable parameters
+        # cannot be determined by TensorFlow
+        assert circuit.qtape.trainable_params == set()
+
+        with tf.GradientTape() as tape:
+            res = circuit(a)
+
+        assert circuit.interface == "tf"
+
+        # with the interface, the tape returns tensorflow tensors
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == tuple()
+
+        # the tape is able to deduce trainable parameters
+        assert circuit.qtape.trainable_params == {0}
+
+        # gradients should work
+        grad = tape.gradient(res, a)
+        assert isinstance(grad, tf.Tensor)
+        assert grad.shape == tuple()
+
+    def test_interface_swap(self, dev_name, diff_method, mode, tol):
+        """Test that the TF interface can be applied to a QNode
+        with a pre-existing interface"""
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, interface="autograd", diff_method=diff_method, mode=mode)
+        def circuit(a):
+            qml.RY(a, wires=0)
+            qml.RX(0.2, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        from pennylane import numpy as anp
+
+        a = anp.array(0.1, requires_grad=True)
+
+        res1 = circuit(a)
+        grad_fn = qml.grad(circuit)
+        grad1 = grad_fn(a)
+
+        # switch to TF interface
+        circuit.interface = "tf"
+
+        a = tf.Variable(0.1, dtype=tf.float64)
+
+        with tf.GradientTape() as tape:
+            res2 = circuit(a)
+
+        grad2 = tape.gradient(res2, a)
+        assert np.allclose(res1, res2, atol=tol, rtol=0)
+        assert np.allclose(grad1, grad2, atol=tol, rtol=0)
+
+    def test_drawing(self, dev_name, diff_method, mode):

```

Comment:
Why is there a drawing test in here?

Reply:
because there was a bug in the circuit drawer a long time ago that didn't allow it to draw operations that had TF tensors 😆 It must be a relic from before we had `qml.draw()`